aws/s3.go

```diff
@@ -88,55 +97,131 @@ func getAllS3Buckets(awsSession *session.Session, excludeAfter time.Time, bucket
 	var bucketNamesPerRegion = make(map[string][]*string)
 	var regionClients = make(map[string]*s3.S3)
 
-	for _, bucket := range output.Buckets {
-		logging.Logger.Debugf("Checking - Bucket %s", *bucket.Name)
+	for _, region := range targetRegions {
+		logging.Logger.Debugf("S3 - creating session - region %s", region)
+		awsSession, err := session.NewSession(&awsgo.Config{
+			Region: awsgo.String(region)},
+		)
+		if err != nil {
+			return bucketNamesPerRegion, errors.WithStackTrace(err)
+		}
+		regionClients[region] = s3.New(awsSession)
+	}
+
+	totalBuckets := len(output.Buckets)
+	if totalBuckets == 0 {
+		return bucketNamesPerRegion, nil
+	}
+
+	if batchSize <= 0 {
+		return nil, fmt.Errorf("Invalid batchsize - %d - should be > 0", batchSize)
+	}
+
+	totalBatches := totalBuckets / batchSize
+	if totalBuckets%batchSize != 0 {
+		totalBatches = totalBatches + 1
+	}
+	batchCount := 1
+
+	// Batch the get operation
+	for i := 0; i < totalBuckets; i += batchSize {
+		j := i + batchSize
+		if j > totalBuckets {
+			j = totalBuckets
+		}
+
+		logging.Logger.Infof("Getting - %d-%d buckets of batch %d/%d", i+1, j, batchCount, totalBatches)
+
+		targetBuckets := output.Buckets[i:j]
+
+		var bucketCh = make(chan *S3Bucket, len(targetBuckets))
+		var wg sync.WaitGroup
 
-		if len(bucketNameSubStr) > 0 {
-			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+		for _, bucket := range targetBuckets {
+			if len(bucketNameSubStr) > 0 && !strings.Contains(*bucket.Name, bucketNameSubStr) {
 				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
 				continue
 			}
-		}
 
-		if !excludeAfter.After(*bucket.CreationDate) {
-			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
-			continue
+			wg.Add(1)
+			go func(bucket *s3.Bucket) {
+				defer wg.Done()
+				if err := getBucketInfo(svc, bucket, regionClients, bucketCh); err != nil {
+					logging.Logger.Warnf("Skipping - Bucket %s - Failed to get info - %s", *bucket.Name, err)
+				}
+			}(bucket)
 		}
 
-		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
-		if err != nil {
-			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
-			continue
-		}
+		go func() {
+			wg.Wait()
+			close(bucketCh)
+		}()
```

Comment:
Why do this in a separate goroutine? Don't you usually call `wg.Wait()` in the original calling goroutine so that routine waits for all the others to finish?

Reply:
I guess you're reading from the channel below, but is there any need to _read_ from the channel concurrently? Seems like the code would be easier to reason about if you structure it to:

1. Start N goroutines to fetch bucket information concurrently
1. When all N goroutines are done, read the data they fetched from the channel

Reply:
Reading from the channel with data as it comes helps in showing messages earlier for skipped buckets when debug mode is on - as opposed to logging everything when everything is done.

Reply:
Roger. Please add that as a comment!

Reply:
Done.